### PR TITLE
Add `prefer-template` rule

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [X] Addresses an existing issue: #772
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [X] Documentation update

#### Overview of change:

Added the `prefer-template` rule, which warns on string literal concatenation.

#### CHANGELOG.md entry:

[new-rule] `prefer-template`

### CODE REVIEW

**Comment by andy-hanson** [comment]
Test failures fixed by #2234.

**Comment by andy-hanson** [comment]
IMO it's always better to use a template (except for simple cases like `x + "!"`).

```ts
const str = "a" +
	someVeryLongFunction(someVeryLongArgument()) +
	"b";
```

should be:

```ts
const x = someVeryLongFunction(someVeryLongArgument());
const str = `a${x}b`;
```

If you like complex nested expressions, you can still do this:

```ts
const str = `a${
	someVeryLongFunction(someVeryLongArgument())
}b`;
```

**Comment by ajafff** [comment]
You're absolutely right for the cases you mentioned.
I meant this:
```ts
const str = "I'm a really really really long text " +
            "that needs to be split into several lines " +
            "to keep it readable. "+
            "But I shall not contain any line breaks. " + 
            "How would you write this as a template string?";
```

**Comment by andy-hanson** [comment]
Actually, it already would never warn for concatenation of string literal + string literal. Added a warning to use a template literal if the strings *did* contain newlines.

File: test/rules/prefer-template/default/test.ts.lint [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+x`a` + y; // OK, can't simplify something with a tag
+
+"a" + "b"; // OK to concatenate strings without newlines.
+
+"a\n" + "b";
+~~~~~~~~~~~ [1]
+
+"a" + "\nb";
+~~~~~~~~~~~ [1]
+
```
Comment (nchen63) [inline test/rules/prefer-template/default/test.ts.lint]: also add test cases for 1+1, 1+1+"a", and x+y

Comment (andy-hanson) [inline test/rules/prefer-template/default/test.ts.lint]: :+1:

**Comment by nchen63** [comment]
@andy-hanson thanks!

**Comment by zakhenry** [comment]
@ajafff not sure which is cleaner, but in your example you can always do
```ts
const str = `I'm a really really really long text\
that needs to be split into several lines\
to keep it readable.\
But I shall not contain any line breaks.\
How would you write this as a template string?"`;
```

**Pull request merged**